src/extension/src/coreTemplateStudio.ts

```diff
@@ -94,57 +99,77 @@ export class CoreTemplateStudio {
   // then it will always get triggered when a command is write to the cli and there is responses from cli, until the process gets killed
   public async readStream(process: ChildProcess) {
     let data = "";
-    process.stdout.on("data", (chunk) => {
+    process.stdout.on("data", chunk => {
       data += chunk;
       const responses = data.toString().split("\n");
       for (let i = 0; i < responses.length - 1; i++) {
         const result = JSON.parse(responses[i]);
-        this.cliEvents.emit(result['messageType'], result['content']);
+        this.cliEvents.emit(result["messageType"], result["content"]);
       }
       data = responses[responses.length - 1];
     });
 
-    process.stderr.on("data", (data) => {
-      this.cliEvents.emit("error", data.toString());
+    process.stderr.on("data", data => {
+      this.cliEvents.emit("eventError", data.toString());
     });
-    process.on("exit", (code) => this.cliEvents.emit("error", `process exited with code ${code}`));
+    process.on("exit", code =>
+      this.cliEvents.emit("eventError", `process exited with code ${code}`)
+    );
   }
 
-  private async awaitCliEvent(eventName: string): Promise<any> {
-    return new Promise((resolve, reject) => {
-      this.cliEvents.once(eventName, (data) => {
-        this.cliEvents.removeAllListeners();
-        resolve(data);
-      }).once("error", (data) => {
-        this.cliEvents.removeAllListeners();
-        reject(data);
+  private async awaitCliEvent(
+    eventName: string,
+    command: string
+  ): Promise<any> {
+    this.promiseChain = this.promiseChain.then(() => {
+      this._processCli.stdin.write(command);
+      return new Promise((resolve, reject) => {
+        this.cliEvents
+          .once(eventName, data => {
+            this.cliEvents.removeAllListeners();
+            resolve(data);
+          })
+          .once("eventError", data => {
+            this.cliEvents.removeAllListeners();
+            reject(data);
+          });
       });
     });
+    return this.promiseChain;
   }
 
   public async sync(payload: ICommandPayload): Promise<any> {
-    const syncCommand = `${CONSTANTS.CLI.SYNC_COMMAND_PREFIX} -p ${payload.payload.path}\n`;
-    this._processCli.stdin.write(syncCommand);
-    this.cliEvents.on(CONSTANTS.CLI.SYNC_PROGRESS_STATE, (data) => {
-      payload.liveMessageHandler(data['status'], data['progress']);
+    const syncCommand = `${CONSTANTS.CLI.SYNC_COMMAND_PREFIX} -p ${
+      payload.payload.path
+    }\n`;
+    this.cliEvents.on(CONSTANTS.CLI.SYNC_PROGRESS_STATE, data => {
```

Comment:
This part also has to be in the promise chain

Reply:
As we talked, we can leave this as is for now :)